storage/store.go

```diff
@@ -1271,33 +1280,78 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
 			RangeLookup: args.Method() == proto.InternalRangeLookup,
 		}
 		bArgs.Add(pushArgs)
+		pushReply := &proto.InternalPushTxnResponse{}
+
+		// Check the txn cache first.
+		if s.pushedTxnCache.processPushedTxn(pushArgs, pushReply) {
+
+			cachedPushTxns = append(cachedPushTxns, &cachedPushTxn{
+				key:   intent.Key,
+				reply: pushReply,
+			})
+
+		} else {
+			bArgs.Add(pushArgs)
+			needPushIntents = append(needPushIntents, intent)
+		}
+
 	}
 	b := &client.Batch{}
 	b.InternalAddCall(client.Call{Args: bArgs, Reply: bReply})
 
-	// Run all pushes in parallel.
-	if pushErr := s.kvDB.Run(b); pushErr != nil {
-		if log.V(1) {
-			log.Infoc(ctx, "push for write intent %s failed: %s", wiErr, pushErr)
-		}
+	if len(needPushIntents) != 0 {
+		b := &client.Batch{}
+		b.InternalAddCall(client.Call{Args: bArgs, Reply: bReply})
 
-		// For write/write conflicts within a transaction, propagate the
-		// push failure, not the original write intent error. The push
-		// failure will instruct the client to restart the transaction
-		// with a backoff.
-		if args.Header().Txn != nil && proto.IsWrite(args) {
-			return pushErr
+		// Run all pushes in parallel.
+		if pushErr := s.kvDB.Run(b); pushErr != nil {
+			if log.V(1) {
+				log.Infoc(ctx, "push for write intent %s failed: %s", wiErr, pushErr)
+			}
+
+			// For write/write conflicts within a transaction, propagate the
+			// push failure, not the original write intent error. The push
+			// failure will instruct the client to restart the transaction
+			// with a backoff.
+			if args.Header().Txn != nil && proto.IsWrite(args) {
+				return pushErr
+			}
+			// For read/write conflicts, return the write intent error which
+			// engages backoff/retry (with !Resolved). We don't need to
+			// restart the txn, only resend the read with a backoff.
+			return wiErr
 		}
-		// For read/write conflicts, return the write intent error which
-		// engages backoff/retry (with !Resolved). We don't need to
-		// restart the txn, only resend the read with a backoff.
-		return wiErr
 	}
 	wiErr.Resolved = true // success!
 
 	// We pushed the transaction(s) successfully, so resolve the intent(s).
-	for i, intent := range wiErr.Intents {
+	for i, cache := range cachedPushTxns {
```

Comment:
This loop and the next one are nearly identical. Can we combine them somehow? Maybe process `needPushIntents` first, updating the cache and `cachedPushTxns`, then issue all the resolve intent calls based on `cachedPushTxns`.